Specifications/QIR/Callables.md

```diff
@@ -205,9 +217,80 @@ The following snippet of LLVM code could be generated:
+  times the function has been applied must be tracked. The `Controlled` functor
+  is not its own inverse.
+  Applying this function to a `%Callable` whose implementation table has a null
+  "ctl" entry should cause a runtime failure.
+- The order of applying these two functions does not need to be tracked. The
+  `Adjoint` and `Controlled` functors commute.
+
+When `__quantum__rt__callable_invoke` is called, the entry in the callable's
+implementation table to be used is selected as follows:
+
+- If the adjoint parity is even and the controlled count is zero, the "body"
+  entry (index 0) should be used.
+- If the adjoint parity is odd and the controlled count is zero, the "adj"
+  entry (index 1) should be used.
+- If the adjoint parity is even and the controlled count is greater than zero,
+  the "ctl" entry (index 2) should be used.
+- If the adjoint parity is odd and the controlled count is greater than zero,
+  the "ctladj" entry (index 3) should be used.
+
+If the controlled count is greater than one, then `__quantum__rt__callable_invoke`
+also needs to do some manipulation of the input tuple.
+Each application of the `Controlled` functor modifies the signature of the
+specialization by adding replacing the current argument tuple with a two-tuple
+containing the array of control qubits as the first element and a tuple of
+the remaining arguments as the second tuple.
+
+For instance, if the base callable expects an argument tuple
+`{ %TupleHeader, i64, %Qubit* }`, then the `Controlled` version expects
+`{ %TupleHeader, %Array*, { %TupleHeader, i64, %Qubit* }* }`, and the twice-`Controlled`
+version expects
+`{ %TupleHeader, %Array*, { %TupleHeader, %Array*, { %TupleHeader, i64, %Qubit* }* }* )`.
+The "ctl" implementation function always expects
+`{ %TupleHeader, %Array*, { %TupleHeader, i64, %Qubit* }* }`.
+Thus, if the controlled count is greater than 1,
+`__quantum__rt__callable_invoke` needs to disassemble the argument tuple,
+concatenate the control qubit arrays, and form the expected argument tuple.
+
+One additional complexity is that the above is modified slightly if the base
+callable expects an argument tuple with exactly one element.
+In this case, the `Controlled` version expects a two-element tuples as
+above, but with the actual base argument as the second element.
+For instance, if the base callable expects `{ %TupleHeader, %Qubit* }`,
+the singly-`Controlled` version expects `{ %TupleHeader, %Array*, %Qubit* }`
+rather than `{ %TupleHeader, %Array*, { %TupleHeader, %Qubit* }* }`.
+This means that the second element of the singly-`Controlled` argument tuple
+is not always a pointer to a struct, and in particular may have variable length
+up to the size of a `%Range`.
+
+To resolve this, `__quantum__rt__callable_invoke` needs to have access to the
+length of the inner argument tuple once it has unwrapped down to that point.
+This could be stored in the `%TupleHeader` by `__quantum__rt__tuple_create`,
```

Comment:
Currently, tuple header is prescribed to be `%TupleHeader = type { i32 }`, where the i32 is used to store reference counting.

Reply:
There are 4 different solutions that come to mind (and I'm sure there are others as well):

1. An implementation could use the bottom 24 bits of the `i32` as the reference count and the high 8 bits as the size in bytes; the size is never greater than 24 (for a `%Range`). Admittedly this is kind of yucky.
1. An implementation could use the `%TupleHeader` i32 to store the size of the tuple and have the `tuple_create` runtime function allocate an extra `i32` at the end of the memory block to hold the actual reference count.
1. An implementation could always allocate an extra 'i32' at the start of the memory block when creating a tuple, store the tuple data size there, and return a pointer past that `i32` from `tuple_create`. When the tuple's reference count goes to zero, the pointer to be freed would have to be calculated by adjusting the pointer to the tuple.
1. An implementation could run an LLVM pass that redefines the `%TupleHeader` structure to something such as `{ i16, i8 }` before further compilation.

Do you think it would be helpful to include some or all of these in the specification?

Reply:
The solutions are standard enough for most people to be aware of them, my concern was that, as worded, it might confuse readers into thinking that %TupleHeader is extensible. It might help to add into the section on %TupleHeader that the implementers are free to use the i32-sized memory however they please (that is, QIR will never try to interpret this i32). Only 8 bits for reference counting seems too low, though. What are the arguments against extending %TupleHeader to {i32, i32} with explicit expectation that one int is reference count and the other is size, or keeping it unspecified but increasing to i64? Solutions 2&3 would use as much memory anyway.